### PR TITLE
Don't show articles from blocked users in the feed

### PR DESCRIPTION
**Pull request by atsmith813**
## What type of PR is this? (check all applicable)
- [x] Bug Fix

## Description
This PR fixes a bug in the functionality of blocking a user. When you block a user, you shouldn't see their posts in your feed. A bug was causing posts from blocked users to show up in people's feeds.

This probably resulted from some refactorings, but the issue seems to be we were using the wrong element identifier. [Articles are rendered in an `article` component](https://github.com/forem/forem/blob/ba972b2c1d726dabd10445afcbae94432094ab5b/app/javascript/articles/Article.jsx#L46-L52), not a `div`.

## Related Tickets & Documents
closes https://github.com/forem/forem/issues/5342
closes https://github.com/forem/forem/issues/11636

## QA Instructions, Screenshots, Recordings
1. Create a new user locally.
2. Fire up the app.
3. Sign in as the local admin (`admin@forem.local` / `password`).
4. Make a new post.
5. Sign out.
6. Sign in as a different user (i.e. from step 1).
7. Make sure you see the post you made in step 4 in the feed (can go to http://localhost:3000/latest as well).
8. Block the admin user/the author of the post created in step 4.
9. Refresh the feed and see that the post is now gone.

## Added tests?
- [x] I need help with writing tests

## [Forem core team only] How will this change be communicated?
- [x] I will share this change internally with the appropriate teams

## [optional] Are there any post deployment tasks we need to perform?
I'm not sure how this JS update will behave in production with caching 😬 

scooby_doo_hiding

### CODE REVIEW

**Review by Zhao-Andy — APPROVED** [review]
me right now

two dudes saying bro, growing more incredulous each time

**Comment by citizen428** [comment]
@atsmith813 Great find! Is there a regression test we can add for this?

**Review by rhymes — COMMENTED** [review]
Great catch!

I wonder how we can make sure this doesn't regress again, should we add an e2e test?

Perhaps it's best to consult the @forem/frontend team before doing that, I don't know if it can be done "easily" with Cypress

**Comment by atsmith813** [comment]
@citizen428 @rhymes yeah I'm not sure the best avenue to approach this as far as testing 🤔 . If I'm being honest, I'd like to avoid an E2E test 😅  but if that's ultimately what's best, I can look into it.

Are we moving away from specs with Capybara? If not, maybe we can test the Article element is visible, block the user, then test the same article element is no longer visible?

Any ideas @forem/frontend?

**Comment by rhymes** [comment]
ccing @aitchiss and @nickytonline to see if they can chime in

**Comment by aitchiss** [comment]
> @citizen428 @rhymes yeah I'm not sure the best avenue to approach this as far as testing 🤔 . If I'm being honest, I'd like to avoid an E2E test 😅 but if that's ultimately what's best, I can look into it.
> 
> _This ticket was originally in the backend column...it just so happened to be a frontend bug_ 🙈
> 
> Are we moving away from specs with Capybara? If not, maybe we can test the Article element is visible, block the user, then test the same article element is no longer visible?
> 
> Any ideas @forem/frontend?

I think this does make sense for a Cypress test (sorry!). I think moving away from Capybara is the plan - Cypress helps us test way closer to genuine user interactions.

We have a helper command to create an article, so in the test setup we could create an article under one user's login, before logging in as another. Then blocking/checking the feed.

If you need a hand with it, I'm happy to assist!

**Comment by aitchiss** [comment]
@atsmith813  - After looking a bit more at the area we'd need to test, I think adding a Cypress test right now isn't actually going to be as straightforward like I thought. 

At the moment, if you try to click either "Flag user" or "Block user" in Cypress from the profile page, the related alert doesn't pop up. I think it's related to this old issue with things not being fully initialised in test - #13228

**TL;DR I think this whole area will need a bit more effort to test robustly with Cypress. We should aim to do that, but I think it should probably be a separate effort to this PR.**

Let's go with a Capybara spec for now just so we know we've tested the feature, and we can test the flag/block flows with Cypress in a separate PR.

**Comment by nickytonline** [comment]
> @atsmith813 - After looking a bit more at the area we'd need to test, I think adding a Cypress test right now isn't actually going to be as straightforward like I thought.
> 
> At the moment, if you try to click either "Flag user" or "Block user" in Cypress from the profile page, the related alert doesn't pop up. I think it's related to this old issue with things not being fully initialised in test - #13228
> 
> **TL;DR I think this whole area will need a bit more effort to test robustly with Cypress. We should aim to do that, but I think it should probably be a separate effort to this PR.**
> 
> Let's go with a Capybara spec for now just so we know we've tested the feature, and we can test the flag/block flows with Cypress in a separate PR.

I wonder if the alert doesn't pop up because maybe the job to generate the alert doesn't run because nothing is ever actually committed to the DB since we are in a transaction that rollsback after every test?

File: spec/services/articles/feeds/basic_spec.rb [hunk 1]
```diff
@@ -45,5 +45,15 @@
       result = feed.feed
       expect(result).not_to include(hot_story)
     end
+
+    it "doesn't display blocked articles", type: :system, js: true do
```
Comment (atsmith813) [inline spec/services/articles/feeds/basic_spec.rb:49]: This could probably live somewhere else 🤔  What do we think?

File: app/javascript/contentDisplayPolicy/hideBlockedContent.js [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 export function hideBlockedContent() {
   const contentUserElements = Array.from(
-    document.querySelectorAll('div[data-content-user-id]'),
+    document.querySelectorAll('article[data-content-user-id]'),
```
Comment (jacobherrington) [inline app/javascript/contentDisplayPolicy/hideBlockedContent.js:5]: :exclamation:

**Review by jacobherrington — APPROVED** [review]
Nice catch! Thanks for adding the regression test :smile:

**Comment by jacobherrington** [comment]
Re: Cypress tests, I agree that it makes sense to do this in an E2E test, but I also think that might make sense as a second PR since this bug is immediately affecting people using Forem.

**Comment by nickytonline** [comment]
> Re: Cypress tests, I agree that it makes sense to do this in an E2E test, but I also think that might make sense as a second PR since this bug is immediately affecting people using Forem.

@jacobherrington, I understand that it's immediately affecting folks, so go ahead and merge, but let's work on the tests right away so they're not forgotten. 😎

**Review by djuber — APPROVED** [review]

**Comment by citizen428** [comment]
@atsmith813 

> If I'm being honest, I'd like to avoid an E2E test

Now saying you have to do it for this PR, but that's how I felt in the past, now I quite enjoy writing them, give Cypress a chance 😄
